parsers/openapi.c

```diff
@@ -0,0 +1,363 @@
+/*
+*
+*   Copyright (c) 2016, Masatake YAMATO
+*   Copyright (c) 2016, Red Hat, K.K.
+*   Copyright (c) 2022, Vasily Kulikov
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   Documentation on the schemas:
+*   https://github.com/OAI/OpenAPI-Specification/tree/main/versions
+*/
+
+#include "general.h"	/* must always come first */
+#include "debug.h"
+#include "entry.h"
+#include "kind.h"
+#include "yaml.h"
+#include "parse.h"
+#include "subparser.h"
+
+#include <stdio.h>
+
+typedef enum {
+	KIND_SCHEMA,
+	KIND_PATH,
+	KIND_RESPONSE,
+	KIND_PARAMETER,
+} openapiKind;
+
+static kindDefinition OpenApiKinds [] = {
+	{ true, 'd', "schema", "schemas" },
+	{ true, 'p', "path", "paths" },
+	{ true, 'R', "response", "responses" },
+	{ true, 'P', "parameter", "parameters" },
+};
+
+enum openapiKeys {
+	KEY_UNKNOWN,
+	KEY_PATHS,
+	KEY_COMPONENTS,
+	KEY_SCHEMAS,
+	KEY_PARAMETERS,
+	KEY_RESPONSES,
+	KEY_DEFINITIONS,
+};
+
+struct yamlBlockTypeStack {
+	yaml_token_type_t type;
+	enum openapiKeys key;
+	struct yamlBlockTypeStack *next;
+};
+
+/* - name: "THE NAME" */
+enum openapiPlayDetectingState {
+	DSTAT_LAST_KEY,
+	DSTAT_LAST_VALUE,
+	DSTAT_INITIAL,
+};
+
+
+struct sOpenApiSubparser {
+	yamlSubparser yaml;
+	struct yamlBlockTypeStack *type_stack;
+	enum openapiPlayDetectingState play_detection_state;
+};
+
+static void pushBlockType (struct sOpenApiSubparser *openapi, yaml_token_type_t t)
+{
+	struct yamlBlockTypeStack *s;
+
+	s = xMalloc (1, struct yamlBlockTypeStack);
+
+	s->next = openapi->type_stack;
+	openapi->type_stack = s;
+
+	s->type = t;
+	s->key = KEY_UNKNOWN;
+}
+
+static void popBlockType (struct sOpenApiSubparser *openapi,
+						  yaml_token_t *token)
+{
+	struct yamlBlockTypeStack *s;
+
+	s = openapi->type_stack;
+	openapi->type_stack = s->next;
+
+	s->next = NULL;
+
+	eFree (s);
+}
+
+static void popAllBlockType (struct sOpenApiSubparser *openapi,
+							 yaml_token_t *token)
+{
+	while (openapi->type_stack)
+		popBlockType (openapi, token);
+}
+
+static bool stateStackMatch (struct yamlBlockTypeStack *stack,
+							 const enum openapiKeys *expectation,
+							 unsigned int length
+							 )
+{
+	if (length == 0)
+	{
+		if (stack == NULL)
+			return true;
+		else
+			return false;
+	}
+
+	if (stack == NULL)
+		return false;
+
+	if (stack->key == expectation[0])
+		return stateStackMatch (stack->next, expectation + 1, length - 1);
+	else
+		return false;
+}
+
+static bool scalarNeq (yaml_token_t *token, unsigned int len, const char* val)
+{
+	if ((token->data.scalar.length == len)
+		&& (strncmp (val, (char *)token->data.scalar.value, len) == 0))
+		return true;
+	else
+		return false;
+}
+
+static enum openapiKeys parseKey(yaml_token_t *token)
+{
+  if (scalarNeq(token, 10, "components"))
```

Comment:
For mapping from "string" to a key, you can use functions defined in main/keyword.h.
I write this here is not for performance. With the function, the code of this parser can be more declarative.